src/enum_maps/side_controller.hpp

```diff
@@ -0,0 +1,49 @@
+/*
+	Copyright (C) 2008 - 2021
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#pragma once
+
+#include <array>
+#include <string>
+
+class side_controller
+{
+public:
+	enum TYPE { INVALID = 0, NONE = 1, HUMAN = 2, AI = 3, RESERVED = 4 };
+	static constexpr const char* const invalid = "invalid";
+	static constexpr const char* const none = "null";
+	static constexpr const char* const human = "human";
+	static constexpr const char* const ai = "ai";
+	static constexpr const char* const reserved = "reserved";
+
+	static const char* get_text(TYPE key)
+	{
+		return values_[key];
+	}
+
+	static TYPE get_enum(const std::string& value)
+	{
+		for(unsigned int i = 0; i < values_.size(); i++)
+		{
+			if(value == values_[i])
+			{
+				return static_cast<TYPE>(i);
+			}
+		}
+		return INVALID;
+	}
+
+private:
+	static constexpr const std::array<const char*, 5> values_ = {invalid, none, human, ai, reserved};
```

Comment:
I feel the MAKE_ENUM macro's technique of putting each pair on the same line is better than splitting them across separate definitions for enum, strings and mapping array.

Reply:
How would you prefer this be done instead then?